### PR TITLE
Assignment9+10 multithreading - Conner Hanson + Trevor Stewart

### PR DESCRIPTION
**Pull request by connerhanson**
Partner @Tvor369

### CODE REVIEW

File: Assignment.Tests/PingProcessTests.cs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Assignment.Tests;
+
+[TestClass]
+public class PingProcessTests
```
Comment (Tfelchlin13) [inline Assignment.Tests/PingProcessTests.cs:13]: Looks good

File: Assignment.Tests/PingProcessTests.cs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    public void Run_GoogleDotCom_Success()
+    {
+        int exitCode = Sut.Run("-c 4 www.google.com").ExitCode;
+        Assert.AreEqual<int>(0, exitCode);
+    }
+    */  
+    //For some reason this test constinously runs even with linux values
+
+    [TestMethod]
+    public void Run_InvalidAddressOutput_Success()
```
Comment (Tfelchlin13) [inline Assignment.Tests/PingProcessTests.cs:41]: I love how you guys write your tests this is way better than my tests

File: Assignment.Tests/PingProcessTests.cs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    public void Run_CaptureStdOutput_Success()
+    {
+        PingResult result = Sut.Run("-c 4 localhost");
+        AssertValidPingOutput(result);
+    }
+
+    [TestMethod]
+    public void RunTaskAsync_Success()// 1.)
+    {
+        Task<PingResult> testOutput = Sut.RunTaskAsync("-c 4 localhost");
```
Comment (Tfelchlin13) [inline Assignment.Tests/PingProcessTests.cs:63]: Looks great

File: Assignment/PingProcess.cs [hunk 4]
```diff
@@ -1,173 +1,217 @@
...
+        void updateStdOutput(string? line) =>
+            (stdOutStringBuilder??=new StringBuilder()).AppendLine(line);
+        void updateStdError(string? line) => 
+            (stdErrorStringBuilder??=new StringBuilder()).AppendLine(line);
+        Process process = RunProcessInternal(StartInfo, updateStdOutput, updateStdError, default);
+        return new PingResult( process.ExitCode, stdOutStringBuilder?.ToString(), stdErrorStringBuilder?.ToString());
+    }
+
+    // 1.)
+    public Task<PingResult> RunTaskAsync(string hostNameOrAddress)
```
Comment (Tfelchlin13) [inline Assignment/PingProcess.cs:32]: looks good

File: Assignment/PingProcess.cs [hunk 5]
```diff
@@ -1,173 +1,217 @@
...
+    {
+        return new Task<PingResult>(() => Run(hostNameOrAddress));
+    }
+
+    // 2.) and 3?
+    async public Task<PingResult> RunAsync(string hostNameOrAddress, CancellationToken cancellationToken = default)
+    {
```
Comment (Tfelchlin13) [inline Assignment/PingProcess.cs:39]: Looks great

File: Assignment/PingProcess.cs [hunk 6]
```diff
@@ -1,173 +1,217 @@
...
+          throw new TaskCanceledException("Operation was canceled", ex);
+      }
+      catch (AggregateException)
+      {
+        throw;
+      }
+    }
+
+    // 4.)
+    async public Task<PingResult> RunAsync(IEnumerable<string> hostNameOrAddresses, CancellationToken cancellationToken = default)
```
Comment (Tfelchlin13) [inline Assignment/PingProcess.cs:59]: This is nice and looks better than how i did it

File: Assignment/PingProcess.cs [hunk 7]
```diff
@@ -1,173 +1,217 @@
...
+            return task.Result.ExitCode;
+        });
+
+        await Task.WhenAll(all);
+        //int total = all.Aggregate(0, (total, item) => total + item.Result);
+        return new PingResult(total, stringBuilder?.ToString().Trim(), null);
+    }
+
+    // 5.)
+    async public Task<int> RunLongRunningAsync(ProcessStartInfo startInfo, Action<string?>? progressOutput, Action<string?>? progressError, CancellationToken token)
```
Comment (Tfelchlin13) [inline Assignment/PingProcess.cs:83]: Looks great!

File: Directory.Build.props [hunk 8]
```diff
@@ -1,9 +1,10 @@
...
-    <Nullable>enable</Nullable>
-    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
-    <EnableNETAnalyzers>true</EnableNETAnalyzers>
-    <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
-  </PropertyGroup>
-</Project>
\ No newline at end of file
+<Project>
+<PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
```
Comment (Tfelchlin13) [inline Directory.Build.props:3]: Looks good

**Review by Tfelchlin13 — COMMENTED** [review]
Overall your code looks very clean and i didnt see any issues with it i love how you write tests they look very clean! 👍

Implement PingProcess' public Task<PingResult> RunTaskAsync(string hostNameOrAddress) ✔
First implement public void RunTaskAsync_Success() test method to test PingProcess.RunTaskAsync() using "localhost". ✔
Do NOT use async/await in this implementation. ✔
Implement PingProcess' async public Task<PingResult> RunAsync(string hostNameOrAddress) ✔
First implement the public void RunAsync_UsingTaskReturn_Success() test method to test PingProcess.RunAsync() using "localhost" without using async/await. ✔
Also implement the async public Task RunAsync_UsingTpl_Success() test method to test PingProcess.RunAsync() using "localhost" but this time DO using async/await. ✔
Add support for an optional cancellation token to the PingProcess.RunAsync() signature. ❌✔ Inside the PingProcess.RunAsync() invoke the token's ThrowIfCancellationRequested() method so an exception is thrown. ✔ Test that, when cancelled from the test method, the exception thrown is an AggregateException ✔ with a TaskCanceledException inner exception ✔ using the test methods RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrapping ✔and RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrappingTaskCanceledException ✔ respectively.
Complete/fix AND test async public Task<PingResult> RunAsync(IEnumerable<string> hostNameOrAddresses, CancellationToken cancellationToken = default) which executes ping for an array of hostNameOrAddresses (which can all be "localhost") in parallel, adding synchronization if needed. ✔ NOTE:
The order of the items in the stdOutput is irrelevant and expected to be intermingled.
StdOutput must have all the ping output returned (no lines can be missing) even though intermingled. ✔
Implement AND test public Task<int> RunLongRunningAsync(ProcessStartInfo startInfo, Action<string?>? progressOutput, Action<string?>? progressError, CancellationToken token) using Task.Factory.StartNew() and invoking RunProcessInternal with a TaskCreation value of TaskCreationOptions.LongRunning and a TaskScheduler value of TaskScheduler.Current. Returning a Task<PingResult> is also okay. NOTE: This method does NOT use Task.Run.
Extra Credit
Test and implement PingProcess.RunAsync(System.IProgress<T> progess) so that you can capture the output as it occurs rather than capturing the output only when the process finishes. ❌ (Not there yet)
Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

File: Assignment.Tests/PingProcessTests.cs [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+      }
+      catch (AggregateException ex)
+      {
+        Exception? flattenedException = ex.Flatten().InnerException;
+
+        Assert.IsInstanceOfType(flattenedException, typeof(TaskCanceledException));
+
+        throw flattenedException!;
+      }
+    }
```
Comment (omar5o9) [inline Assignment.Tests/PingProcessTests.cs:118]: Good job on this test

**Review by omar5o9 — COMMENTED** [review]
Implement PingProcess' public Task<PingResult> RunTaskAsync(string hostNameOrAddress) ✔
First implement public void RunTaskAsync_Success() test method to test PingProcess.RunTaskAsync() using "localhost". ✔
Do NOT use async/await in this implementation. ✔
Implement PingProcess' async public Task<PingResult> RunAsync(string hostNameOrAddress) ✔
First implement the public void RunAsync_UsingTaskReturn_Success() test method to test PingProcess.RunAsync() using "localhost" without using async/await. ✔
Also implement the async public Task RunAsync_UsingTpl_Success() test method to test PingProcess.RunAsync() using "localhost" but this time DO using async/await. ✔
Add support for an optional cancellation token to the PingProcess.RunAsync() signature. ✔ Inside the PingProcess.RunAsync() invoke the token's ThrowIfCancellationRequested() method so an exception is thrown. ✔ Test that, when cancelled from the test method, the exception thrown is an AggregateException ✔ with a TaskCanceledException inner exception ✔ using the test methods RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrapping ✔and RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrappingTaskCanceledException ✔ respectively.
Complete/fix AND test async public Task<PingResult> RunAsync(IEnumerable<string> hostNameOrAddresses, CancellationToken cancellationToken = default) which executes ping for an array of hostNameOrAddresses (which can all be "localhost") in parallel, adding synchronization if needed. ✔ NOTE:
The order of the items in the stdOutput is irrelevant and expected to be intermingled.
StdOutput must have all the ping output returned (no lines can be missing) even though intermingled. ✔
Implement AND test public Task<int> RunLongRunningAsync(ProcessStartInfo startInfo, Action<string?>? progressOutput, Action<string?>? progressError, CancellationToken token) using Task.Factory.StartNew() and invoking RunProcessInternal with a TaskCreation value of TaskCreationOptions.LongRunning and a TaskScheduler value of TaskScheduler.Current. Returning a Task<PingResult> is also okay. NOTE: This method does NOT use Task.Run.
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

Good job overall

File: Assignment.Tests/PingProcessTests.cs (hunk 9 above)
Comment (connerhanson) [reply Assignment.Tests/PingProcessTests.cs:118]: Thank you!

File: Assignment.Tests/PingProcessTests.cs (hunk 2 above)
Comment (connerhanson) [reply Assignment.Tests/PingProcessTests.cs:41]: Thank you, it was a tricky one.

File: Assignment/PingProcess.cs (hunk 6 above)
Comment (connerhanson) [reply Assignment/PingProcess.cs:59]: Thank you!

File: Assignment.Tests/PingProcessTests.cs [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+
+    [TestMethod]//Re-named to a more appopriate name
+    public void Run_GetPingReply_Success()
+    {
+        //GitHub can't handle pinging google apparently so will ping localhost instead
+        string host = (Environment.GetEnvironmentVariable("GITHUB_ACTIONS") == null) ? 
+            "-c 4 google.com" : "-c 4 localhost";
+        int exitCode = Sut.Run(host).ExitCode;
+        Assert.AreEqual(0, exitCode);
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/PingProcessTests.cs:38]: Nice

File: Assignment.Tests/PingProcessTests.cs [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+    public void Run_InvalidAddressOutput_Success()
+    {
+        (int exitCode, string? stdOutput, string? stdError) = Sut.Run("badaddress");
+        Assert.IsFalse(string.IsNullOrWhiteSpace(stdError));
+        stdOutput = WildcardPattern.NormalizeLineEndings(stdError!.Trim());
+        Assert.AreEqual<string?>(
+            "ping: badaddress: Temporary failure in name resolution".Trim(),
+            stdOutput,
+            $"Output is unexpected: {stdOutput}");
+        Assert.AreEqual<int>(2, exitCode); //In linux bad address results in exit code 2 not 1
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/PingProcessTests.cs:51]:
```suggestion
        Assert.NotEqual<int>(0, exitCode);
```

File: Assignment.Tests/PingProcessTests.cs [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+        PingResult result = Sut.Run("-c 4 localhost");
+        AssertValidPingOutput(result);
+    }
+
+    [TestMethod]
+    public void RunTaskAsync_Success()// 1.)
+    {
+        Task<PingResult> testOutput = Sut.RunTaskAsync("-c 4 localhost");
+        testOutput.Start();//Need to schedule so it can complete
+        AssertValidPingOutput(testOutput.Result);
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/PingProcessTests.cs:66]: Isn't the task already hot here and `.Start` not needed? 

https://essentialcsharp.com/asynchronous-tasks#introducing-asynchronous-tasks

File: Assignment.Tests/PingProcessTests.cs [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+    [TestMethod]
+    async public Task RunAsync_MultipleHostAddresses_True()// 4.)
+    {
+        // Pseudo Code - don't trust it!!! <-- I'm trustin it!
+        string[] hostNames = ["-c 4 localhost", "-c 4 localhost", "-c 4 localhost", "-c 4 localhost"];
+        int expectedLineCount = PingOutputLikeExpression.Split(Environment.NewLine).Length * hostNames.Length;
+        PingResult result = await Sut.RunAsync(hostNames);
+        int? lineCount = result.StdOutput?.Split(Environment.NewLine).Length;
+        Assert.AreEqual(expectedLineCount, lineCount);
+        //Assert.AreEqual("", result.StdOutput);
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/PingProcessTests.cs:130]:
```suggestion
PingResult pingResult = await Sut.RunAsync("localhost");
        string[] hostNames = new string[] { "localhost", "localhost", "localhost", "localhost" };
        int expectedLineCount = PingOutputLikeExpression.Split(Environment.NewLine).Length*hostNames.Length;
        int? expectedLineCount = pingResult.StdOutput?.Split(Environment.NewLine).Length*hostNames.Length;
        PingResult result = await Sut.RunAsync(hostNames);
        int? lineCount = result.StdOutput?.Split(Environment.NewLine).Length;
        Assert.AreEqual(expectedLineCount, lineCount);
```
Don't trust it :)

File: Assignment/Assignment.csproj [hunk 14]
```diff
@@ -0,0 +1,8 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+	<PropertyGroup>
+    <LangVersion>12.0</LangVersion>
+    <Nullable>enable</Nullable>
```
Comment (BenjaminMichaelis) [inline Assignment/Assignment.csproj:5]: Why aren't these just in the props file?

File: Assignment.Tests/Assignment.Tests.csproj [hunk 15]
```diff
@@ -1,16 +1,18 @@
...
-	  <Using Include="Microsoft.VisualStudio.TestTools.UnitTesting" />
-  </ItemGroup>
-
-</Project>
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <IsPackable>false</IsPackable>
+    <Nullable>enable</Nullable>
+	  <LangVersion>12.0</LangVersion>
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/Assignment.Tests.csproj:6]: Why aren't these just in the props file?

File: Assignment/PingProcess.cs [hunk 16]
```diff
@@ -1,173 +1,204 @@
...
+    {
+      var task = Task.Factory.StartNew(() =>
+      {
+        Process process = RunProcessInternal(startInfo, progressOutput, progressError, token);
```
Comment (BenjaminMichaelis) [inline Assignment/PingProcess.cs:87]:
```suggestion
      {
      cancellationToken.ThrowIfCancellationRequested();
        Process process = RunProcessInternal(startInfo, progressOutput, progressError, token);
```

File: Assignment/PingProcess.cs [hunk 17]
```diff
@@ -1,173 +1,204 @@
...
+            return task.Result.ExitCode;
+        });
+
+        await Task.WhenAll(all);
+        //int total = all.Aggregate(0, (total, item) => total + item.Result);
+        return new PingResult(total, stringBuilder?.ToString().Trim(), null);
+    }
+
+    // 5.)
+    async public Task<int> RunLongRunningAsync(ProcessStartInfo startInfo, Action<string?>? progressOutput, Action<string?>? progressError, CancellationToken token)
```
Comment (BenjaminMichaelis) [inline Assignment/PingProcess.cs:83]: Why change this method signature?

**Pull request closed without merging**